enhancements/api-review/custom-tags-aws.md

```diff
@@ -73,71 +95,400 @@ Note existing unchanged behavior: The installer will apply these tags to all AWS
+
+#### Delete tags scenarios
+Tags are deleted when the user sets the user-defined tag value to an empty string in `.spec.platformSpec.aws.resourceTags`.
+Also refer to `Precedence` scenario to understand cases where deletion of user-defined tags is not allowed .
+
+For example,
+
+Existing tag for AWS resource = `key_infra1 = value_old`
+
+New tag request = `.spec.platformSpec.aws.resourceTags` has `key_infra1 =`
+
+Action = Existing tag for AWS resource is deleted.
+
+Final tag set to AWS resource = deleted
+
+Event action = An event is generated to notify user about the request status (success/failure) to delete tags for the AWS resource.
+
+#### Precedence scenarios
+1) User-defined tags on kube resources MUST continue to take precedence during create and update. User-defined tags found on kube resources must not be deleted by methods described in `Delete tags scenarios`.
+   A warning is generated to inform the user about action not being applied on the list of user tags.
+
+   For example,
+
+   Existing tag in kube resource object like machine CR = `key_infra1 = custom_value`
+
+   New tag request = `.spec.platformSpec.aws.resourceTags` has `key_infra1 = value1`
+
+   Action = The tag for AWS resource is maintained to `key_infra1 = custom_value`.
+
+   Final tag set to AWS resource = `key_infra1 = custom_value`
+
+   Event action = An event is generated to notify user about the request status (failure) to update tags for the AWS resource. A warning also must be generated about list of user-defined tags on which action is not applicable.
+
+2) When a new user-defined tag is added on kube resources, it MUST override the tag in `.spec.platformSpec.aws`. User-defined tags found on kube resources must not be deleted by methods described in `Delete tags scenarios`.
+   A warning is generated to inform the user about action not being applied on the list of user tags.
+
+   For example,
+
+   Existing tag in Infrastructure CR = `key_infra1 = custom_value`
+
+   New tag added in kube resource object like machine CR = `key_infra1 = value1`
+
+   Action = The tag for AWS resource is updated to `key_infra1 = value1`.
+
+   Final tag set to AWS resource = `key_infra1 = value1`
+
+   Event action = An event is generated to notify user about the request status (success) to update tags for the AWS resource. A warning also must be generated about list of user-defined tags on which action did a override.
+
+3) `.spec.platformSpec.aws` take precedence over `.status.platformStatus.aws`. User-defined tags must be merged from `.status.platformStatus.aws` and set to AWS resource.
+
+4) Before deleting user-defined tag from AWS resource, the entry must be removed from `.status.platformStatus.aws`, if any, as operator will reconcile user-defined tag to the AWS resource.
+   Setting user-defined tag to empty string in `.spec.platformSpec.aws` will not override the user-defined tag in `.status.platformStatus.aws`.
```

Comment:
The value in spec takes precedence only for updates. The status is maintained immutable in the upgrades also.
A central infra controller to sync spec to status field. The controller becomes a SPOF. As the different operators do not update request status back to the central controller, avoiding SPOF is better. As the status field is set for deprecation, future revisions of the feature may not require this function.

Reply:
Why would the value in spec only take precedence on updates?

Reply:
As delete is not supported for status. Only updates are considered for precedence.

Reply:
I don't follow how delete not being supported for status is relevant. The precedence applies to the current values in the spec and not how the actions performed to get to the current values.

Reply:
You are right, precedence is not related to actions. 

> The value in spec takes precedence only for updates. The status is maintained immutable in the upgrades also. A central infra controller to sync spec to status field becomes a SPOF. As the different operators do not update request status back to the central controller, avoiding SPOF is better. The status field is set for deprecation, future revisions of the feature may not require this function.

The meaning in the conversation here is, in case of delete operations, status value is not updated. Tags in spec override tags in status when there is an update operation. Delete operation fails if there is a tag of same key is found in status.

Reply:
> The status is maintained immutable in the upgrades also.
> 
> A central infra controller to sync spec to status field becomes a SPOF. As the different operators do not update request status back to the central controller, avoiding SPOF is better. The status field is set for deprecation, future revisions of the feature may not require this function.

If the tag from status is not removed, on the next reconcile operator will add the tag entry back to AWS resource.

Is it ok to go ahead with above limitations about status field and allow delete when there is tag with same key in status?

Reply:
The scenario being discussed here is the one where the user sets the value to an empty string. It is not the scenario where the tag is removed from `resourceTags`. The next reconcile will still see that the spec is directing the operator to delete the tag from the AWS resource, regardless of what is in the infra status.